include/hipSYCL/sycl/libkernel/host/group_functions.hpp

```diff
@@ -0,0 +1,514 @@
+        result = false;
+        break;
+      }
+    }
+  }
+  return group_broadcast(g, result);
+}
+
+template <typename Group, typename Ptr, typename Predicate>
+bool none_of(Group g, Ptr first, Ptr last, Predicate pred) {
+  bool result = true;
+  auto lid = g.get_local_linear_id();
+
+  if (lid == 0) {
+    while (first != last) {
+      if (pred(*(first++))) {
+        result = false;
+        break;
+      }
+    }
+  }
+  return group_broadcast(g, result);
+}
+
+
+// reduce
+template <typename Group, typename V, typename T, typename BinaryOperation>
+T reduce(Group g, V *first, V *last, T init, BinaryOperation binary_op) {
+  T result{};
+  auto lid = g.get_local_linear_id();
+
+  if (first >= last) {
+    return init;
+  }
+
+  if (lid == 0) {
+    result = binary_op(*(first++), init);
+    while (first != last)
+      result = binary_op(result, *(first++));
+  }
+  return group_broadcast(g, result);
+}
+
+template <typename Group, typename T, typename BinaryOperation>
+T reduce(Group g, T *first, T *last, BinaryOperation binary_op) {
+  return reduce(g, first, last, T{}, binary_op);
+}
+
+
+// exclusive_scan
+template <typename Group, typename V, typename T, typename BinaryOperation>
+T *exclusive_scan(Group g, V *first, V *last, T *result, T init, BinaryOperation binary_op) {
+  auto lid = g.get_local_linear_id();
+
+  if (lid == 0) {
+    *(result++) = init;
+    while (first != last - 1) {
+      *result = binary_op(*(result - 1), *(first++));
+      result++;
+    }
+  }
+  return group_broadcast(g, result);
+}
+
+template <typename Group, typename V, typename T, typename BinaryOperation>
+T *exclusive_scan(Group g, V *first, V *last, T *result, BinaryOperation binary_op) {
+  return exclusive_scan(g, first, last, result, T{}, binary_op);
+}
+
+
+// inclusive_scan
+template <typename Group, typename V, typename T, typename BinaryOperation>
+T *inclusive_scan(Group g, V *first, V *last, T *result, T init, BinaryOperation binary_op) {
+  if (first == last)
+    return result;
+
+  auto lid = g.get_local_linear_id();
+
+  if (lid == 0) {
+    *(result++) = binary_op(init, *(first++));
+    while (first != last) {
+      *result = binary_op(*(result - 1), *(first++));
+      result++;
+    }
+  }
+  return group_broadcast(g, result);
+}
+
+template <typename Group, typename V, typename T, typename BinaryOperation>
+T *inclusive_scan(Group g, V *first, V *last, T *result, BinaryOperation binary_op) {
+  return inclusive_scan(g, first, last, result, T{}, binary_op);
+}
+
+
+} // namespace detail
+
+// broadcast
+template <typename Group, typename T>
+HIPSYCL_KERNEL_TARGET
+T group_broadcast(Group g, T x, typename Group::linear_id_type local_linear_id = 0) {
+  T *scratch = detail::get_shared_memory_ptr<T>(g);
+  auto lid = g.get_local_linear_id();
+
+  if (lid == local_linear_id) {
+    scratch[0] = x;
+  }
+
+  group_barrier(g);
+
+  T tmp = scratch[0];
+
+  group_barrier(g);
+
+  detail::free_shared_ptr(g, scratch);
```

Comment:
We are in SYCL. We call it local memory, except when referring explicitly to HIP/CUDA and their `__shared__` memory :-)